lochness/mindlamp/__init__.py

```diff
@@ -38,46 +92,95 @@ def sync(Lochness: 'lochness.config',
+    days_to_check = get_days_to_pull(Lochness)
+
+    # current time (ct) in UTC
+    ct_utc = datetime.now(pytz.timezone('UTC'))
+
+    # set the cut off point as UTC 00:00:00.00
+    ct_utc_00 = ct_utc.replace(hour=0, minute=0, second=0, microsecond=0)
+
+    for days_from_ct in range(days_to_check):
+        # n day before current time
+        time_utc_00 = ct_utc_00 - timedelta(days=days_from_ct)
+        time_utc_00_ts = time.mktime(time_utc_00.timetuple()) * 1000
+        time_utc_24 = time_utc_00 + timedelta(hours=24)
+        time_utc_24_ts = time.mktime(time_utc_24.timetuple()) * 1000
+
+        # date string to be used in the file name
+        date_str = time_utc_00.strftime("%Y_%m_%d")
+
+        # Extra information for future version
+        # study_id, study_name = get_study_lamp(LAMP)
+        # subject_ids = get_participants_lamp(LAMP, study_id)
+        subject_id = subject.mindlamp[f'mindlamp.{subject.study}'][0]
+        logger.debug(f'Mindlamp {subject_id} {date_str} data pull - start')
+
+        # pull data from mindlamp
+        begin = time.time()
+        activity_dicts = get_activity_events_lamp(
+                LAMP, subject_id,
+                from_ts=time_utc_00_ts, to_ts=time_utc_24_ts)
+
+        sensor_dicts = get_sensor_events_lamp(
+                LAMP, subject_id,
+                from_ts=time_utc_00_ts, to_ts=time_utc_24_ts)
+        end = time.time()
+        logger.debug(f'Mindlamp {subject_id} {date_str} data pull - complete')
+
+        # set destination folder
+        # dst_folder = tree.get('mindlamp', subject.general_folder)
+        dst_folder = tree.get('mindlamp',
+                              subject.protected_folder,
+                              processed=False,
+                              BIDS=Lochness['BIDS'])
+
+        # store both data types
+        for data_name, data_dict in zip(['activity', 'sensor'],
+                                        [activity_dicts, sensor_dicts]):
+
+            dst = os.path.join(
+                    dst_folder,
+                    f'{subject_id}_{subject.study}_{data_name}_'
+                    f'{date_str}.json')
+
+            # separate out audio data from the activity dictionary
+            if data_name == 'activity' and data_dict != []:
+                sound_dst = os.path.join(
+                        dst_folder,
+                        f'{subject_id}_{subject.study}_{data_name}_'
+                        f'{date_str}_sound.mp3')
+                data_dict = get_audio_out_from_content(data_dict, sound_dst)
+
+            jsonData = json.dumps(
+                data_dict,
+                sort_keys=True, indent='\t', separators=(',', ': '))
+
+            content = jsonData.encode()
+
+            if content.strip() == b'[]':
+                logger.info(f'No mindlamp data for {subject_id} {date_str}')
+                continue
+
+
+            if not Path(dst).is_file():
```

Comment:
Why are we doing this extra work? Does `atomic_write` take a long time? Why don't we just overwrite silently?

Even though `atomic_write` takes a long time, comparison against existing json + overwriting should take more time!

Reply:
This is checking if the file has already been downloaded by a previous run of `sync.py`. If there is no file, execute the download. If there is already the existing file, we compare the files using `crc32` algorithm to make sure we are not missing any data from the server. (redownload if the existing file does not match the new data, and skip downloading if match)

(`json.dump` does not save a file, if this was what you're pointing to)

Reply:
Instead of downloading and comparing, why don't we just download and overwrite?

Reply:
Good point. I tried to avoid unnecessary overwriting. Would you prefer to overwrite each time?

Reply:
Yes. To reiterate:

> comparison against existing json + overwriting should take more time!

Reply:
removed - ade5fc9
We may want to revisit in future, if we need `lochness.backup` function when there is a conflict in the existing data vs new data.